// p2pspider/lib/btclient.js
'use strict';

exports.__esModule = true;

var _net = require('net');

var _net2 = _interopRequireDefault(_net);

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

var _lru = require('lru');

var _lru2 = _interopRequireDefault(_lru);

var _wire = require('./wire');

var _wire2 = _interopRequireDefault(_wire);

var _utils = require('./utils');

var _utils2 = _interopRequireDefault(_utils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var BTClient = function (_EventEmitter) {
  _inherits(BTClient, _EventEmitter);

  /**
   * constructor
   * @param  {Object} options [description]
   * @return {[type]}         [description]
   */

  function BTClient() {
    var options = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, BTClient);

    var _this = _possibleConstructorReturn(this, _EventEmitter.call(this));

    _this.timeout = options.timeout || 5000;
    _this.lru = (0, _lru2.default)({
      max: 100000,
      maxAge: 1000 * 60 * 10
    });
    return _this;
  }
  /**
   * format meta data
   * @param  {[type]} metadata [description]
   * @return {[type]}          [description]
   */


  BTClient.prototype.formatMetaData = function formatMetaData(metadata) {
    var info = metadata.info;
    var name = info['utf-8.name'] || info.name;
    if (!name) {
      return;
    }
    name = _utils2.default.toUtf8String(name);

    var data = {
      name: name,
      size: info.length
    };
    if (info.private) {
      data.private = info.private;
    }
    if (info.files) {
      var total = 0;
      data.files = info.files.map(function (item) {
        item.path = item.path.map(function (it) {
          return _utils2.default.toUtf8String(it);
        }).join('/');
        total += item.length;
        return {
          size: item.length,
          path: item.path
        };
      }).sort(function (a, b) {
        return a.size > b.size ? -1 : 1;
      });
      data.size = total;
    } else {
      data.files = [{
        size: data.size,
        path: data.name
      }];
    }

    var extraProperties = ['source', 'profiles', 'private', 'file-duration', 'file-media', 'pieces'];
    extraProperties.forEach(function (item) {
      if (info[item]) {
        data[item] = info[item];
      }
    });
	data.torrentFileData = metadata;

    return data;
  };
  /**
   * download
   * @param  {Object} rinfo    [description]
   * @param  {[type]} infohash [description]
   * @return {[type]}          [description]
   */


  BTClient.prototype.download = function download() {
    var _this2 = this;

    var rinfo = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var infohash = arguments[1];

    var infoHashHex = infohash.toString('hex');
    if (this.lru.get(infoHashHex)) {
      return;
    }
    this.lru.set(infoHashHex, true);

    var socket = new _net2.default.Socket();

    socket.setTimeout(this.timeout);

    socket.connect(rinfo.port, rinfo.address, function () {
      var wire = new _wire2.default(infohash);
      socket.pipe(wire).pipe(socket);
      wire.on('metadata', function (metadata, infoHash) {
        //destroy socket when get metadata
        socket.destroy();

        metadata = _this2.formatMetaData(metadata);
        if (!metadata) {
          return;
        }
        _this2.emit('complete', metadata, infoHash, rinfo);
      });
      wire.sendHandshake();
    });

    socket.on('error', function () {
      socket.destroy();
    });

    socket.on('timeout', function () {
      socket.destroy();
    });
  };

  return BTClient;
}(_events2.default);

exports.default = BTClient;
